File: 3036-number-of-subarrays-that-match-a-pattern-ii/3036-number-of-subarrays-that-match-a-pattern-ii.cpp
class Solution {
public:
    vector<int> lcs(vector<int>& s) {
        int n = s.size();
        vector<int> dp(n, 0);
        dp[0] = 0;
        for(int i = 1; i < n; ++i) {
            int j = dp[i - 1];
            while(j > 0 && s[i] != s[j]) j = dp[j-1];
            dp[i] = j + (s[i] == s[j]);
        }
        return dp;
    };
    
    int countMatchingSubarrays(vector<int>& s, vector<int>& pattern) {
        for(int i=0;i<s.size()-1;i++)
        {
            int a = s[i];
            int b = s[i+1];
            
            if(a<b)
                s[i] = 1;
            if(a == b)
                s[i] = 0;
            if(a>b)
                s[i] = -1;
           
        } //convert s[i] to {-1,0,1}
        vector<int> ns(pattern);
        ns.push_back(4);
        for(int x : s) ns.push_back(x);
        vector<int> dp = lcs(ns);
        int count = 0;
        for(int i = 0; i < dp.size(); ++i)
            if(dp[i] == pattern.size()) ++count;
        return count;
    }
};

File: 3036-number-of-subarrays-that-match-a-pattern-ii/3036-number-of-subarrays-that-match-a-pattern-ii.py
class Solution(object):
    def countMatchingSubarrays(self, nums, pattern):
        """
        :type nums: List[int]
        :type pattern: List[int]
        :rtype: int
        """
        def lcs(s):
            n = len(s)
            dp = [0]*n
            for i in range(1, n):
                j = dp[i-1]
                while j > 0 and s[i] != s[j]:
                    j = dp[j - 1]
                dp[i] = j + (s[i] == s[j])
            return dp
        
        arr = map(cmp, nums[1:], nums[:-1])
        ns = lcs(pattern + [256] + arr)
        return ns.count(len(pattern))